torch/_dynamo/cache_size.py

```diff
@@ -0,0 +1,154 @@
+entries on the forward code object. In the older case when we had only 1 limit,
+our cache size limit must be >= 32 to capture all these recompilations. Now,
+suppose there is a separate function in the same program which is very dynamic
+and unsuitable for compilation. Such a function will need to undergo 32
+compilations to burst the cache and fallback to eager. These 32 recompilations
+are too many and we want to fallback for these compilation-unfriendly functions
+sooner.
+
+In the new scenario, we can have (1) cache_size_limit = 2, (2)
+accumulated_cache_size_limit = 32. This means that each ID_MATCH'd object can
+have maximum of two cache entries, and the maximum number of cache entries
+(irrespective of ID_MATCH obj) is 32. This covers the case of forward code
+object which has 32 recompilations. For the other function, the one unsuitable
+for recompilation, our limit is 2. So, we will burst the cache in just 2
+recompilations. In this manner, these 2 limits help us resolve the tension
+mentioned earlier.
+"""
+
+
+@dataclass
+class CacheSize:
+    """
+    Tracks the different cache size limits per code object. See the note at the
+    top of the file.
+    """
+
+    # Total number of CacheEntry objects in the Dynamo linked list
+    total: int = 0
+
+    # Number of CacheEntry objects for an object with ID_MATCH guard.
+    per_id_guarded_obj: DefaultDict[ReferenceType, int] = field(
+        default_factory=defaultdict(int)
+    )
+
+    # Sources of objects with ID_MATCH guards
+    id_guarded_sources: Set[str] = field(default_factory=set)
+
+
+def compute_cache_size(frame: types.FrameType, cache_entry) -> CacheSize:
+    # Walk the linked list to calculate the cache size
+    cache_size_per_id_matched_obj = defaultdict(int)
+    total = 0
+    sources = set()
+    while cache_entry:
+        total += 1
+        for source, weak_id in cache_entry.check_fn.id_matched_objs.items():
+            if weak_id() is not None:
+                cache_size_per_id_matched_obj[weak_id] += 1
+                sources.add(source)
+        cache_entry = cache_entry.next
+
+    return CacheSize(total, cache_size_per_id_matched_obj, sources)
+
+
+def _get_weakref_from_f_locals(frame: types.FrameType, source: str):
+    obj = frame.f_locals.get(source, None)
+    weak_id = None
+    try:
+        weak_id = weakref.ref(obj)
+    except TypeError:
+        pass  # cannot weakref bool object
+    return weak_id
+
+
+def is_recompilation(frame: types.FrameType, cache_size: CacheSize) -> bool:
+    sources = cache_size.id_guarded_sources
+
+    if not sources:
+        return cache_size.total >= 1
+
+    per_id_guarded_obj = cache_size.per_id_guarded_obj
+    for source in sources:
+        weak_id = _get_weakref_from_f_locals(frame, source)
+        if (
+            weak_id
+            and weak_id in per_id_guarded_obj
+            and per_id_guarded_obj[weak_id] >= 1
+        ):
+            return True
```

Comment:
This is interesting. Here, you are saying that it counts as a recompile whenever there is an ID_MATCH that we've already seen before. But let's suppose we have a setup like this:

```
some_random_global_that_stays_same = object()
nn_module1 = Module()
nn_module2 = Module()

nn_module1.forward(...)
nn_module2.forward(...)
```

IMO, we should not count the second forward as a recompile. But if Module.forward references `some_random_global_that_stays_same` in a way that triggers an ID_MATCH, your logic here would count it as a recompile, because we have two compiles attached to `some_random_global_that_stays_same`.

It seems to me that it should only count as a recompile if ALL of the guarded objects >=1.

Reply:
@ezyang I changed the logic of what is called a recompilation. Also revolved the logic around a `__ge__` on CacheSize to easily reason about is_recompilation and exceeds_cache_size_limit.